flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java

```diff
@@ -309,22 +354,83 @@ public void testScheduleCallableWithDelayInSeconds() throws Exception {
 
         final MainThreadExecutable mainThreadExecutable =
-                new TestMainThreadExecutable(
-                        (runnable, delay) -> actualDelayMsFuture.complete(delay));
+                new TestMainThreadExecutable((runnable) -> taskCompletedFuture.complete(null));
 
         final RpcEndpoint.MainThreadExecutor mainThreadExecutor =
-                new RpcEndpoint.MainThreadExecutor(mainThreadExecutable, () -> {});
+                new RpcEndpoint.MainThreadExecutor(mainThreadExecutable, () -> {}, endpointId);
 
         final Duration expectedDelay = Duration.ofSeconds(1);
 
         scheduler.accept(mainThreadExecutor, expectedDelay);
 
-        assertThat(actualDelayMsFuture.get(), is(expectedDelay.toMillis()));
+        taskCompletedFuture.get();
+        mainThreadExecutor.close();
+    }
+
+    private static void testScheduleAfterClose(
+            BiFunction<RpcEndpoint.MainThreadExecutor, Duration, ScheduledFuture<?>> scheduler) {
+        final CompletableFuture<Void> taskCompletedFuture = new CompletableFuture<>();
+        final String endpointId = "foobar";
+
+        final Duration expectedDelay = Duration.ofSeconds(1);
+        final MainThreadExecutable mainThreadExecutable =
+                new TestMainThreadExecutable((runnable) -> taskCompletedFuture.complete(null));
+
+        final RpcEndpoint.MainThreadExecutor mainThreadExecutor =
+                new RpcEndpoint.MainThreadExecutor(mainThreadExecutable, () -> {}, endpointId);
+
+        mainThreadExecutor.close();
+        ScheduledFuture<?> future = scheduler.apply(mainThreadExecutor, expectedDelay);
+
+        assertTrue(future.isDone());
+        assertTrue(future instanceof ThrowingScheduledFuture);
+        assertTrue(future.cancel(true));
+        assertTrue(future.cancel(false));
+        assertFalse(taskCompletedFuture.isDone());
+    }
+
+    private static void testCancelScheduledTask(
+            BiFunction<RpcEndpoint.MainThreadExecutor, Duration, ScheduledFuture<?>> scheduler) {
+        final CompletableFuture<Void> actualDelayMsFuture = new CompletableFuture<>();
+        final String endpointId = "foobar";
+
+        final MainThreadExecutable mainThreadExecutable =
+                new TestMainThreadExecutable((runnable) -> actualDelayMsFuture.complete(null));
+
+        final RpcEndpoint.MainThreadExecutor mainThreadExecutor =
+                new RpcEndpoint.MainThreadExecutor(mainThreadExecutable, () -> {}, endpointId);
+        final Duration timeDelay = Duration.ofSeconds(10);
+        ScheduledFuture<?> scheduledFuture = scheduler.apply(mainThreadExecutor, timeDelay);
+
+        assertEquals(1, mainThreadExecutor.getMainScheduledExecutor().getQueue().size());
+        scheduledFuture.cancel(true);
+        mainThreadExecutor.close();
+
+        assertTrue(scheduledFuture.isCancelled());
+        assertTrue(mainThreadExecutor.getMainScheduledExecutor().getQueue().isEmpty());
+        assertFalse(actualDelayMsFuture.isDone());
```

Comment:
I see two problems in this test:

1. We should verify that `(runnable) -> actualDelayMsFuture.complete(null)` will not be executed, rather than verifying it's not passed to `mainThreadExecutor.runAsync()`. Actually, canceling `scheduledFuture` will not stop it from being scheduled to `mainThreadExecutor`. It is `FutureTask` that guarantees itself won't be executed once canceled.

Currently, the test will fail if you remove `mainThreadExecutor.close()`, because it is a `() -> {}` or `() -> 1` that actually wrapped int the `scheduledFuture`, thus cancelling it will not stop `actualDelayMsFuture` from being completed.

2. Another challenge here is to make sure task is not executed before being canceled.
- If we schedule with a short delay, the test might be unstable because the task can finish before canceling.
- If we schedule with a large delay, we would need to wait longer to see if the task is performed, if not looking into the internal queue of the executor.

I would suggest to use a short delay, while use `assumeTrue` to skip the test if the task is occasionally finished before canceling. You can find out whether the task has been finished before canceling from the return value of `cancel()`.